segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3Storage.java

```diff
@@ -0,0 +1,380 @@
+        }
+
+        try (InputStream reader = client.readObjectStream(config.getBucket(),
+                config.getRoot() + handle.getSegmentName(), Range.fromOffsetLength(offset, length))) {
+
+            if (reader == null) {
+                throw new StreamSegmentNotExistsException(handle.getSegmentName());
+            }
+
+            int bytesRead = StreamHelpers.readAll(reader, buffer, bufferOffset, length);
+
+            LoggerHelpers.traceLeave(log, "read", traceId, bytesRead);
+            return length;
+        }
+    }
+
+    private StreamSegmentInformation syncGetStreamSegmentInfo(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "getStreamSegmentInfo", streamSegmentName);
+        S3ObjectMetadata result = client.getObjectMetadata(config.getBucket(),
+                config.getRoot() + streamSegmentName);
+
+        AccessControlList acls = client.getObjectAcl(config.getBucket(), config.getRoot() + streamSegmentName);
+
+        boolean canWrite = acls.getGrants().stream().anyMatch((grant) -> grant.getPermission().compareTo(Permission.WRITE) >= 0);
+
+        StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName,
+                result.getContentLength(), !canWrite, false,
+                new ImmutableDate(result.getLastModified().toInstant().toEpochMilli()));
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName);
+        return information;
+    }
+
+    private boolean syncExists(String streamSegmentName) {
+        ListObjectsResult result = null;
+        try {
+            result = client.listObjects(config.getBucket(), config.getRoot() + streamSegmentName);
+            return !result.getObjects().isEmpty();
+        } catch (S3Exception e) {
+            if ( e.getErrorCode().equals("NoSuchKey")) {
+                return false;
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    private SegmentProperties syncCreate(String streamSegmentName) throws StreamSegmentExistsException {
+        long traceId = LoggerHelpers.traceEnter(log, "create", streamSegmentName);
+
+        if (!client.listObjects(config.getBucket(), config.getRoot() + streamSegmentName).getObjects().isEmpty()) {
+            throw new StreamSegmentExistsException(streamSegmentName);
+        }
+
+        S3ObjectMetadata metadata = new S3ObjectMetadata();
+        metadata.setContentLength((long) 0);
+
+        PutObjectRequest request = new PutObjectRequest(config.getBucket(),
+                config.getRoot() + streamSegmentName,
+                (Object) null);
+
+        AccessControlList acl = new AccessControlList();
+        acl.addGrants(new Grant[]{
+                new Grant(new CanonicalUser(config.getAccessKey(), config.getAccessKey()),
+                        Permission.FULL_CONTROL)
+        });
+
+        request.setAcl(acl);
+
+        client.putObject(request);
```

Comment:
1. In the scenario Tom mentioned. The second call to putObject can not know that an object already exists. From AWS putobject:
```
Amazon S3 is a distributed system. If it receives multiple write requests for the same object simultaneously, it overwrites all but the last object written. Amazon S3 does not provide object locking; if you need this, make sure to build it into your application layer.
```
2. It is still a non issue. The second host that picks up the create operation now owns the container. Which means it will own subsequent write operations also. So in this scenario there is no loss of data.